dias/tracker.py

```diff
@@ -0,0 +1,418 @@
+        for batch in chunked(records, 100):
+            FileTable.insert_many(
+                batch,
+                fields=[
+                    FileTable.filepath,
+                    FileTable.file_type_id,
+                    FileTable.start_time,
+                    FileTable.end_time,
+                    FileTable.exists,
+                ],
+            ).on_conflict(action="IGNORE").execute()
+            # on_conflict specifies that we do not insert duplicates
+
+    def remove_files(self, files):
+        """
+        Set exists = False for files in FileTable.
+
+        Parameters
+        ----------
+        files : List of String
+            list of filepaths
+        """
+        print(f"File Tracker: Settings exists = False for {len(files)} in database")
+
+        with db.atomic():
+            for f in files:
+                file_row = FileTable.get(FileTable.filepath == f)
+                file_row.exists = False
+                file_row.save()
+
+    def update_file_table(self):
+        """Insert novel files that are in `self.path` into db, and set exists=False for files that are not at `self.path`, anymore."""
+        files_on_disk = []
+        for ft in ["chimestack", "chimecal", "chimetiming"]:
+            files_on_disk.extend(
+                glob.glob(os.path.join(self.base_path, f"*_{ft}_corr", "*.h5"))
+            )
+
+        files_in_db = [
+            query.filepath
+            for query in (FileTable.select().where(FileTable.exists == True))
+        ]
+
+        self.insert_files([f for f in files_on_disk if f not in files_in_db])
+        self.remove_files([f for f in files_in_db if f not in files_on_disk])
+
+    def new_files(self, update=True):
+        """
+        Return a list of files unprocessed by self.dias_task_name, of its filetype of interest.
+
+        Optionally updates the File Table with the current state of `self.path`.
+
+        Parameters
+        ----------
+        update : Boolean
+            If true, updates the File Table with the current state of `self.path`.
+        """
+        if update:
+            self.update_file_table()
+
+        # Get file type that the analyzer is interested in
+        # assumes analyzer is interested in only one file type
+        file_type_id = (
+            Analyzer.select(Analyzer, FileType)
+            .where(Analyzer.name == self.dias_task_name)
+            .join(FileType)
+        )[0].file_type_id
+
+        # Select files processed by analyzer
+        files_processed = (
+            FileTable.select()
+            .join(Processed)
+            .join(Analyzer)
+            .where(Processed.analyzer_id.name == self.dias_task_name)
+        )
+
+        # SELECT files of analyzer.file_type_id and NOT Processed by Analyzer
+        # TODO Filter by date
+        files_unprocessed = (
+            FileTable.select()
+            .where(FileTable.id.not_in(files_processed))
```

Comment:
Is there a way to directly put in the query that you want the ones that are not in the `processed` table, in order not to have to query all processed ones first?

Reply:
I had a really tough time with it! I could not get the Join to work out. I tried a Left Outer Join, and then filtered for `None`s. I agree that it would be preferable to not have to query for all of the Processed first. I will give it another shot.

Reply:
I think this is fine. In theory it'll form a sql statement that uses a subquery to do this and it won't actually run two queries. You could check that by doing `files_unprocessed.sql()` to see what it's actually going to run.